### PR TITLE
Use split glob pattern for identifier, force relative path

### PR DESCRIPTION
**Pull request by lukescott**
This PR does the following:

- Moves helper functions such as `makeImport`, `freezeNamespaceObject`, and `makeNamespaceObject` out of the main function body.

- Forces path to be relative, even if `glob:` is used.

- Removes `common-path-prefix` and `common-extname` in favor of breaking apart the glob expression. The unique portion is used.

  `GlobSync` is used instead of `glob.sync` to get more information. `glob.sync` is a helper of `GlobSync` that returns the `result.found`.

  `result.minimatch.set` is the expression split by `/`. Minimatch, glob's matching library, represents literal parts of the expression as a `string`, dynamic parts as a `RegExp`, and `**` as `GLOBSTAR` (ref constant).

  NOTE: Glob supports more than one expression. So `set` is actually a 2-D array. Since 1 expression is only used in this package, the outer array always has a length of 1. So in the code we actually use `set[0]`.

  The min and max part of the split expression is used as unique part of the path:
  - `./foo/*/bar/*/bat.js` resolves to `*/bar/*`
  - `./foo/*/bar/*.js` resolves to `*/bar/*` (if filename is included extension is removed)

### CODE REVIEW

**Comment by lukescott** [comment]
Ok, so I was removing the whole extension using `nodePath.extname`. The issue with this is you could have conflicts if you had something like `foo.one.txt` and `foo.two.txt`. That was the point of `common-extname`.

The issue with `common-extname`, however, is it would remove `.one.txt` if you only had files with`.one.txt`. That would change if you added a file with `.two.txt` into the same directory. It would, at first, remove `.one.js`, then only remove `.js` because you added a file. That can be frustrating.

So instead I'm pulling the extension from the import expression. If you have `*.js` only `.js` is removed. If you have `*.one.js` then `.one.js` is removed. If you add a file with `.two.js` that file isn't included, so the behavior doesn't change.

I came across a similar problem with `common-path-prefix` with the older version. I had an include that looked like `./*/*.reducer.js`. I only had `./textField/textField.reducer.js`. The identifier was `textField`. As soon as I added `multipleChoice/multipleChoice.reducer.js` the first one became `textField$textField`.

With this PR and the extname changes the addition of files will not change the result since the determined identifier will be based on the pattern.

File: README.md
Comment (novemberborn) [inline README.md]: "are also removed"

File: index.js
Comment (novemberborn) [inline index.js]: I don't find this manual destructuring worthwhile, actually. It's not a big deal but if you'd want to change it that be great.

File: index.js
Comment (novemberborn) [inline index.js]: `glob/sync` doesn't seem to be documented. Perhaps access this through `require('glob').GlobSync`? Though I can't see that class in the README either, perhaps use `require('glob').Glob` with the `sync` option?

File: index.js
Comment (novemberborn) [inline index.js]: There can be expressions that result in multiple sets:

```
❯ node -pe "require('glob').Glob('node_modules/{glob,minimatch}/package.json',{sync:true}).minimatch.set"
[ [ 'node_modules', 'glob', 'package.json' ],
  [ 'node_modules', 'minimatch', 'package.json' ] ]
```

File: index.js
Comment (novemberborn) [inline index.js]: What happens if the `glob:` starting pattern doesn't have any magic?

File: index.js
Comment (novemberborn) [inline index.js]: Given the `hasMagic()` check statement above, why is this condition reached?

File: test/import-glob.js
Comment (novemberborn) [inline test/import-glob.js]: This test seems obsolete given the change in behavior, and the new tests added at the end?

File: test/import-glob.js
Comment (novemberborn) [inline test/import-glob.js]: This test seems obsolete given the change in behavior, and the new tests added at the end?

File: index.js
Comment (novemberborn) [inline index.js]: `for (const m of members) {}` instead?

File: index.js
Comment (novemberborn) [inline index.js]: `for/of`

File: index.js
Comment (novemberborn) [inline index.js]: `for/of`

File: index.js
Comment (novemberborn) [inline index.js]:
```js
for (const m of members) {
  replacement.push(makeImport(t, `_${localName}_${m.name}`, m.relative))
}
```

File: index.js
Comment (novemberborn) [inline index.js]: This can only be `true` if `specifiers.length === 0`. Perhaps wrap the above logic in an a `specifiers.length > 0` condition and put this in an `else` branch?

File: index.js
Comment (novemberborn) [inline index.js]: Perhaps use `gm.found` directly?

File: index.js
Comment (novemberborn) [inline index.js]: `subpath` rather than `part`?

File: index.js
Comment (novemberborn) [inline index.js]: The naming of `ext` and `extLen` could be clearer.

File: index.js
Comment (novemberborn) [inline index.js]: Could you add a comment explaining what this is (not) trying to match?

File: index.js
Comment (novemberborn) [inline index.js]: `_src` isn't documented in `minimatch`. `p.source.slice(1, -1)` should be the same though.

File: index.js
Comment (novemberborn) [inline index.js]: I would do something like:

```js
if (p === GLOBSTAR) {

} else if (typeof p === 'string') {

} else {
  // regexp
}
```

File: index.js
Comment (novemberborn) [inline index.js]: `makeSubpathExpression`?

File: index.js
Comment (novemberborn) [inline index.js]: Since you only assign this once you could include the `(` when first pushing the part of the expression.

File: index.js
Comment (novemberborn) [inline index.js]: Could you add a comment explaining this expression?

File: index.js
Comment (novemberborn) [inline index.js]: What does this replacement achieve?

File: index.js
Comment (novemberborn) [reply index.js]: It's also unfortunate that `_glob` isn't documented in `minimatch`.

**Review by novemberborn — CHANGES_REQUESTED** [review]
Wow, this is great!

> the determined identifier will be based on the pattern.

This is a really clever approach, solving the various issues you've raised without awkward syntax or unpredictable behavior 👍 

I've left a whole bunch of comments, some around edge-cases, other about using undocumented properties / functions, some clarifications around what the various regular expressions do and then some stylistic ones (sorry).

I'm excited about seeing this land!

File: index.js
Comment (lukescott) [reply index.js]: The documentation also says:
> sync DEPRECATED: use glob.sync(pattern, opts) instead.

Which means if I apply the same logic to the class, I need to use the `GlobSync` class as well. It does appear that the `GlobSync` class is exported in `require('glob')` as well, so I can change that at least.

File: index.js
Comment (lukescott) [reply index.js]: Restructuring is supported in Node 6. So when Node 4 is dropped it's an easier conversion:

```js
const {
  dirname,
  relative,
  resolve,
  sep as fileSeparator
} = require('path')
```

Also `nodePath` looks strange in the code to me.

File: index.js
Comment (lukescott) [reply index.js]: Both `GLOBSTAR` and `typeof p !== 'string'` are regular expression - GLOBSTAR is just pre-defined. I want to capture min/max on both of those. If I were to do something like above, I would be duplicating that logic.

File: index.js
Comment (lukescott) [reply index.js]: Yeah, I'm assuming it only adds `^` `$`. I'll verify.

File: index.js
Comment (lukescott) [reply index.js]: This escapes the literal string. I had this wrapped in a function, but I only used it once. I could make it a function again. Or comment it. I pulled it from minimatch.

File: index.js
Comment (lukescott) [reply index.js]: I think they provide the glob expressions on `minimatch` separately, similar to set. I'll see if I can use that instead.

File: index.js
Comment (lukescott) [reply index.js]: Hmm. I should be able to get around this by changing  `makeRe`/`makeSubpathExpression` to return an array of regular expressions. I'll need to merge `extLen` into it though. The return would look like: [{regexp: {}, ext: '.js'},...]. Then we just figure out which expression matches each file.

File: index.js
Comment (lukescott) [reply index.js]: I could throw an error.

File: index.js
Comment (lukescott) [reply index.js]: This would be reached if they just had `"glob:"`. If I add a `hasMagic` when `pattern.startsWith('glob:')` is true, then I can throw an error if pattern is empty or isn't a glob pattern.

File: index.js
Comment (lukescott) [reply index.js]: Sure. This is more out of habit. When using babel the transformation of for/of adds a try/catch, which is useless for arrays. Not using babel anymore though :).

File: index.js
Comment (lukescott) [reply index.js]: Sure. My thought on this was to only use one `path.replaceWithMultiple` at the end.

File: test/import-glob.js
Comment (lukescott) [reply test/import-glob.js]: Not sure. I just maintained the existing tests. If the branch coverage stays and same and you feel they are not needed I'll remove them.

File: index.js
Comment (novemberborn) [reply index.js]: OK, let's go with the `GlobSync` export.

File: index.js
Comment (novemberborn) [reply index.js]: I know `nodePath` is odd… perhaps `path` in the `ImportDeclaration` visitor should be renamed `astPath` or `babelPath`.

File: index.js
Comment (novemberborn) [reply index.js]: Fair enough. Perhaps start with `if(typeof p === 'string')` and avoid the negation. Anyway, small fry.

File: index.js
Comment (novemberborn) [reply index.js]: A function would be good, that tends to make it self-documenting. Would be good to reference a specific line in the minimatch source as well.

File: index.js
Comment (novemberborn) [reply index.js]: Yea, let's clean that up above, where the pattern is being normalized.

File: index.js
Comment (novemberborn) [reply index.js]: That's cool, I was trying to highlight that checking `replacement.length` is an indirect test for `specifiers.length === 0`, so we could just have an `if`/`else` condition on that.

File: test/import-glob.js
Comment (novemberborn) [reply test/import-glob.js]: Yea let's try it.

**Comment by lukescott** [comment]
@novemberborn I'm done with the first pass. I mostly focused on adding brace support. Several of the items above are fixed, although with the code changes I'm sure there will be more comments.

I did not remove the two tests yet - I think it may be a good idea to go thru each test and clean it up a bit. I put in a couple brace tests using existing fixtures, which may not make much sense. When I have more time I can go over the tests in more detail.

File: index.js
Comment (novemberborn) [inline index.js]: I suppose this could be `return specifiers.some(s => s.type === 'ImportDefaultSpecifier')`.

File: test/import-glob.js
Comment (novemberborn) [inline test/import-glob.js]: There should be a `three.csv` file here.

**Review by novemberborn — CHANGES_REQUESTED** [review]
Nice! Anything I can help with to get this over the finish line?

**Comment by lukescott** [comment]
I made all the review changes mentioned above. At some point I want to revisit the tests and go over them in more detail, but that shouldn't hold up the merge.

Sorry for the delay. I've been dealing with a family crisis since last week.

**Comment by novemberborn** [comment]
Thanks, I'll try and have a look tomorrow.

> Sorry for the delay. I've been dealing with a family crisis since last week.

No worries, that's far more important. Hope it ends well.

**Comment by novemberborn** [comment]
I've pushed some tweaks, mostly renaming the very short variables which, for me at least, helps with readability.

I've also tried to refactor the `makeSubpathExpressions` function but I'm a bit confused by parts of it. I've added `FIXME` comments for that. Perhaps you could elaborate on those?

I'd like to take a pass through the README to make sure it's phrased in terms of the new glob-determines-name behavior. I haven't looked in depth at the tests yet.

File: index.js
Comment (lukescott) [inline index.js]: If you have `/*/bar.{txt,csv}` then `bar.txt` and `bar.csv` are strings. So they don't get captured, unless we compare between expressions to see that they are different.

File: index.js
Comment (lukescott) [inline index.js]: Same reasoning as above. The filenames are braced, so they are captured. So we want to get the extension. We only grab the extension if the filename is captured.

File: index.js
Comment (lukescott) [inline index.js]: Hmmmm.... I mean we could by using `minimatch.set` instead. I'm just concerned the regexp could change. `minimatch.globParts` is just the brace-expanded glob pattern, which should stay consistent. If we were to use `minimatch.set`, we would have to de-duplicate the expressions.

File: index.js
Comment (lukescott) [inline index.js]: This could possibly be:
```js
pattern = pattern.substr('glob:'.length)
```
`glob:` could be changed to a constant. I think the regexp might be a tad overkill here.

**Comment by lukescott** [comment]
> I've also tried to refactor the makeSubpathExpressions function but I'm a bit confused by parts of it. I've added FIXME comments for that. Perhaps you could elaborate on those?

My earlier comments no longer show. Basically, yes, that code was for matching braced patterns. The logic was basically if one of the other patterns did not match, it was dynamic even if it was a string. Which was only a problem if it was at the end. For exampe,`foo.{csv,txt}` produces `foo.csv` and `foo.txt`.

I refactored the code to stop using `globSet` (not documented anyway) and I match the extension from the RegExp pattern instead. I also place the ending `)` before the extension, so I no longer have to treat extension separately. Also it now creates a single RegExp for all sets.

**Comment by novemberborn** [comment]
I've been at a conference this weekend, will try to have a look at this tomorrow.

Meanwhile, is it OK if I add you as a maintainer on this repo and npm? These are really great contributions.

**Comment by novemberborn** [comment]
@lukescott very nice!

I've pushed a fix for patterns like `{a,b/c}/d`, in which case the `minimatch.set` array contains pattern lists of varying lengths. This was quite tricky to work through so I ended up rewriting a fair bit of your recent code. I'm also using `Set()` to obviate the need for `lodash.uniq`. Let me know what you think.

**Comment by novemberborn** [comment]
(Still need to review tests and update the readme).

**Comment by lukescott** [comment]
Hmmm. I didn't consider `/` inside the brace. What happens with multiple instances of those in different positions? minimatch's makeRe takes separate patterns and combines them with `|`. With flattenSet I was trying to make a shorter pattern. I think it may be easier (and more readable) to revert it to the original `makeSubpathExpressions` and simply join each expression with a `|`.

**Comment by lukescott** [comment]
Hmmm... I didn't consider `/` in path. What happens when you have more than one of these? I'm not sure I fully understand what's going on in the revised code.

**Comment by lukescott** [comment]
Here are a couple interesting patterns:
```
./fixtures/{use-dir-name,}/{one,two,thr/ee}/foo.txt
./fixtures/{use-dir-name,use-dir/name2}/{one,two,thr/ee}/foo.txt
```
Which generates
```
^(?:\.)/(?:fixtures)/((?:use-dir-name|one|two|thr)(?:/(?:one|two|thr|ee))?(?:/(?:ee))?)/(?:foo\.txt)$
^(?:\.)/(?:fixtures)/((?:use-dir-name|use-dir)/(?:one|two|thr|name2)(?:/(?:ee|one|two|thr))?(?:/(?:ee))?)/(?:foo\.txt)$
```
The expressions are less precise. They allow for invalid matches... although glob has already done matching, so maybe it's ok?

minimatch's makeRe makes each expression separately and joins with `|`. It's a longer expression though:

```
/^(?:\.\/fixtures\/use\-dir\-name\/one\/foo\.txt|\.\/fixtures\/use\-dir\-name\/two\/foo\.txt|\.\/fixtures\/use\-dir\-name\/thr\/ee\/foo\.txt|\.\/fixtures\/use\-dir\/name2\/one\/foo\.txt|\.\/fixtures\/use\-dir\/name2\/two\/foo\.txt|\.\/fixtures\/use\-dir\/name2\/thr\/ee\/foo\.txt)$/
/^(?:\.\/fixtures\/use\-dir\-name\/one\/foo\.txt|\.\/fixtures\/use\-dir\-name\/two\/foo\.txt|\.\/fixtures\/use\-dir\-name\/thr\/ee\/foo\.txt|\.\/fixtures\/one\/foo\.txt|\.\/fixtures\/two\/foo\.txt|\.\/fixtures\/thr\/ee\/foo\.txt)$/
```

The reason we don't use minimatch's makeRe is because it lacks capture / extension isolation.

Perhaps we can find a solution in the middle. Identify the captured portion and expand only that part of the expression. So the above would look like this:

```
^\./fixtures/(use-dir-name/one|use-dir-name/two|use-dir-name/thr/ee|one|two|thr/ee)/foo\.txt$
^\./fixtures/(use-dir-name/one|use-dir-name/two|use-dir-name/thr/ee|use-dir/name2/one|use-dir/name2/two|use-dir/name2/thr/ee)/foo\.txt$
```

It might actually be a good idea to split the expression generator off into it's own thing.

**Comment by lukescott** [comment]
I created a package called `minimatch-capture`, which now does most of the heavy lifting for capturing. This package has been slimmed down quite a bit. `minimatch-capture` has 67 tests covering varying use cases.

One change is braces are no longer considered capturable by themselves. For instance, with the path `./fixtures/use-dir-name/{one,two,thr/ee}/foo.txt` there is nothing to capture because it is split into 3 patterns.

In that test I changed it to `./fixtures/use-dir-name/{!(sub),@(sub)/*}/foo.txt`, which captures:
```
./fixtures/use-dir-name/one/foo.txt = one
./fixtures/use-dir-name/two/foo.txt = two
./fixtures/use-dir-name/sub/three/foo.txt = sub/three
./fixtures/use-dir-name/sub/four/foo.txt = sub/four
```
You can change the pattern to `./fixtures/use-dir-name/{!(sub),sub/*}/foo.txt`, which captures:
```
./fixtures/use-dir-name/one/foo.txt = one
./fixtures/use-dir-name/two/foo.txt = two
./fixtures/use-dir-name/sub/three/foo.txt = three
./fixtures/use-dir-name/sub/four/foo.txt = four
```

The reasons for this:

1. The test `./fixtures/ext-from-pattern/*.{txt,csv}` wants to omit the extension. `minimatch-capture` doesn't care about extension.

2. When I tried to make `minimatch-capture`'s `split` method include braces and figure out the left/middle/right before expanding the pattern, it broke a lot of tests. The matching was no longer consistent with minimatch.

**Comment by novemberborn** [comment]
Hey @lukescott, apologies for dropping off the radar. Will try to get back to you on this tomorrow.

**Comment by novemberborn** [comment]
> I created a package called minimatch-capture, which now does most of the heavy lifting for capturing. This package has been slimmed down quite a bit. minimatch-capture has 67 tests covering varying use cases.

Awesome!

> One change is braces are no longer considered capturable by themselves. For instance, with the path `./fixtures/use-dir-name/{one,two,thr/ee}/foo.txt` there is nothing to capture because it is split into 3 patterns.
> 
> In that test I changed it to `./fixtures/use-dir-name/{!(sub),@(sub)/*}/foo.txt`

If I'm understanding this correctly, `*` and `**` are automatically captured. You can use `!(pattern)`, `?(pattern)`, `+(pattern)`, `*(pattern)` and `@(pattern)` to ensure other path components are captured. Brace expansions are not necessarily captured, unless they're between other capture groups. For example `./*/{foo,bar}/*.txt` causes `foo` and `bar` to be included in the identifiers. Otherwise it's not possible to capture brace expansions (`({foo,bar})` doesn't work), so they have to be rewritten using the parentheses-matching syntax.

I've updated the tests and the readme, though we should probably document the capturing behavior better, especially as it comes to brace expansions. Beyond that I think this is good to go!

**Comment by lukescott** [comment]
> If I'm understanding this correctly, * and ** are automatically captured. You can use !(pattern), ?(pattern), +(pattern), *(pattern) and @(pattern) to ensure other path components are captured. Brace expansions are not necessarily captured, unless they're between other capture groups. For example ./*/{foo,bar}/*.txt causes foo and bar to be included in the identifiers. Otherwise it's not possible to capture brace expansions (({foo,bar}) doesn't work), so they have to be rewritten using the parentheses-matching syntax.

Yep, that's correct.

**Review by novemberborn — APPROVED** [review]

**Comment by novemberborn** [comment]
Landed! 🚀 ✨ 🎉  And shipped as 2.0.0!

I've added you to this repo and on npm as well. Many many thanks for pushing this new approach.

**Pull request merged**
